src/java.desktop/share/classes/sun/java2d/pipe/RenderBuffer.java

```diff
@@ -25,57 +25,57 @@
+ * environment.  Its functionality is similar to the ByteBuffer and related
  * NIO classes.  However, the methods in this class perform little to no
  * alignment or bounds checks for performance reasons.  Therefore, it is
  * the caller's responsibility to ensure that all put() calls are properly
  * aligned and within bounds:
  *   - int and float values must be aligned on 4-byte boundaries
  *   - long and double values must be aligned on 8-byte boundaries
+ * Failure to do so will result in exceptions from the FFM API, or worse.
  *
  * This class only includes the bare minimum of methods to support
  * single-threaded rendering.  For example, there is no put(double[]) method
  * because we currently have no need for such a method in the STR classes.
  */
-public class RenderBuffer {
+public final class RenderBuffer {
 
     /**
      * These constants represent the size of various data types (in bytes).
      */
-    protected static final long SIZEOF_BYTE   = 1L;
-    protected static final long SIZEOF_SHORT  = 2L;
-    protected static final long SIZEOF_INT    = 4L;
-    protected static final long SIZEOF_FLOAT  = 4L;
-    protected static final long SIZEOF_LONG   = 8L;
-    protected static final long SIZEOF_DOUBLE = 8L;
+    private static final int SIZEOF_BYTE   = 1;
+    private static final int SIZEOF_SHORT  = 2;
+    private static final int SIZEOF_INT    = 4;
+    private static final int SIZEOF_FLOAT  = 4;
+    private static final int SIZEOF_LONG   = 8;
+    private static final int SIZEOF_DOUBLE = 8;
 
     /**
-     * Represents the number of elements at which we have empirically
-     * determined that the average cost of a JNI call exceeds the expense
-     * of an element by element copy.  In other words, if the number of
-     * elements in an array to be copied exceeds this value, then we should
-     * use the copyFromArray() method to complete the bulk put operation.
-     * (This value can be adjusted if the cost of JNI downcalls is reduced
-     * in a future release.)
+     * Measurements show that using the copy API from a segment backed by a heap
+     * array gets reliably faster than individual puts around a length of 10.
+     * However the time is miniscule in the context of what it is used for
+     * and much more than adequate, so no problem expected if this changes over time.
      */
-    private static final int COPY_FROM_ARRAY_THRESHOLD = 6;
-
-    protected final Unsafe unsafe;
-    protected final long baseAddress;
-    protected final long endAddress;
-    protected long curAddress;
-    protected final int capacity;
-
-    protected RenderBuffer(int numBytes) {
-        unsafe = Unsafe.getUnsafe();
-        curAddress = baseAddress = unsafe.allocateMemory(numBytes);
-        endAddress = baseAddress + numBytes;
+    private static final int COPY_FROM_ARRAY_THRESHOLD = 10;
+
+    MemorySegment segment;
+    private final long baseAddress;
+    private long curOffset;
+    private final int capacity;
```

Comment:
You could perhaps drop the `capacity` field, and use `MemorySegment::byteSize` in it's place.

Reply:
P.S. same for `baseAddress` I think. Looks like it's only used in `getAddress`. Could replace with `segment.address()`.